File: omf-core-framework/src/main/java/com/samares_engineering/omf/omf_core_framework/errormanagement2/logging/OMFLogger2.java
package com.samares_engineering.omf.omf_core_framework.errormanagement2.logging;

import com.nomagic.magicdraw.core.Application;
import com.nomagic.magicdraw.ui.notification.Notification;
import com.nomagic.magicdraw.ui.notification.NotificationManager;
import com.nomagic.magicdraw.ui.notification.NotificationSeverity;
import com.samares_engineering.omf.omf_core_framework.errormanagement2.exceptions.CoreException2;
import com.samares_engineering.omf.omf_core_framework.errormanagement2.exceptions.OMFWarningException;
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.log.OMFLog2;
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.log.OMFLogLevel2;
import com.samares_engineering.omf.omf_core_framework.feature.MDFeature;
import com.samares_engineering.omf.omf_core_framework.plugin.APlugin;
import com.samares_engineering.omf.omf_core_framework.utils.ColorPrinter;

public class OMFLogger2 {
    private static OMFLogger2 instance;
    private APlugin plugin;
    private OMFLogLevel2 logLevel = OMFLogLevel2.INFO;

    private OMFLogger2(APlugin plugin) {
        this.plugin = plugin;
    }

    //Rational, to avoid typing OMFLLogger2.getInstance() everytime, it's included in each static method
    private static OMFLogger2 getInstance() {
        if (instance.plugin == null) {
            throw new CoreException2("The OMFLogger has not been initialized yet. Please call the init() method first.");
        }
        return instance;
    }

    public static void init(APlugin plugin) {
        if (instance != null) {
            throw new CoreException2("Can't initialize the OMFLogger has it has already been initialized.");
        }
        instance = new OMFLogger2(plugin);
    }

    public static void logToUIConsole(OMFLog2 logMessage, OMFLogLevel2 logLevel) {
        if (logLevel.ordinal() >= getInstance().logLevel.ordinal()) { //if the log level is higher than the current log level
            logMessage.replaceNewLinesWithBreaks();
            String formattedLog = logMessage.toHTMLFormat(logLevel, getInstance().plugin.getName());
            Application.getInstance().getGUILog().addHyperlinkedText(formattedLog, logMessage.getLinkActionMapping());
        }
    }

    public static void logToUIConsole(String message, OMFLogLevel2 logLevel) {
        logToUIConsole(new OMFLog2().text(message), logLevel);
    }

    public static void logToUIConsole(OMFLog2 logMessage, OMFLogLevel2 logLevel, MDFeature feature) {
        if (logLevel.ordinal() >= getInstance().logLevel.ordinal()) {
            String formattedLog = logMessage.toHTMLFormat(logLevel, getInstance().plugin.getName(), feature.getName());
            Application.getInstance().getGUILog().addHyperlinkedText(formattedLog, logMessage.getLinkActionMapping());
        }
    }

    public static void logToUIConsole(String message, OMFLogLevel2 logLevel, MDFeature feature) {
        logToUIConsole(new OMFLog2().text(message), logLevel, feature);
    }

    public static void logToNotification(OMFLog2 logMessage, OMFLogLevel2 logLevel, MDFeature feature) {
        if (logLevel.ordinal() >= getInstance().logLevel.ordinal()) {
            NotificationManager.getInstance().showNotification(new Notification(
                    "[Plugin Error]", //id (not sure what is does)
                    OMFLog2.getPrefix(logLevel, getInstance().plugin.getName(), feature.getName()), //title
                    logMessage.replaceNewLinesWithBreaks().toString(),
                    getNotificationSeverity(logLevel))
            );
        }
    }

    public static void logToNotification(OMFLog2 logMessage, OMFLogLevel2 logLevel) {
        if (logLevel.ordinal() >= getInstance().logLevel.ordinal()) {
            NotificationManager.getInstance().showNotification(new Notification(
                    "[Plugin Error]", //id (not sure what is does)
                    OMFLog2.getPrefix(logLevel, getInstance().plugin.getName()), //title
                    logMessage.replaceNewLinesWithBreaks().toString(),
                    getNotificationSeverity(logLevel))
            );
        }
    }

    public static void logToNotification(String message, OMFLogLevel2 logLevel, MDFeature feature) {
            logToNotification(new OMFLog2().text(message).replaceNewLinesWithBreaks(), logLevel, feature);
    }

    public static void logToNotification(String message, OMFLogLevel2 logLevel) {
        logToNotification(message, logLevel, null);
    }

    public static void logToSystemConsole(OMFLog2 logMessage, OMFLogLevel2 logLevel) {
        if (logLevel.ordinal() >= getInstance().logLevel.ordinal()) {
            switch (logLevel) {

                case WARNING:
                    ColorPrinter.warn(logMessage.toString());
                    break;
                case ERROR:
                    ColorPrinter.err(logMessage.toString());
                    break;
                case INFO:
                default:
                    ColorPrinter.print(logMessage.toString());
                    break;
            }
        }
    }

    public static void logToSystemConsole(String message, OMFLogLevel2 logLevel) {
        logToSystemConsole(new OMFLog2().text(message), logLevel);
    }

    public static void logToSystemConsole(OMFLog2 logMessage, OMFLogLevel2 logLevel, MDFeature feature) {
        logToSystemConsole("[" + feature.getName() + "]" + logMessage.toString(), logLevel);
    }

    private static NotificationSeverity getNotificationSeverity(OMFLogLevel2 logLevel) {
        switch (logLevel) {
            case WARNING:
                return NotificationSeverity.WARNING;
            case ERROR:
                return NotificationSeverity.ERROR;
            case INFO:
            default:
                return NotificationSeverity.INFO;
        }
    }

    public static void warnToNotification(String message) {
        logToNotification(message, OMFLogLevel2.WARNING);
    }

    public static void errorToNotification(String message) {
        logToNotification(message, OMFLogLevel2.ERROR);
    }

    public static void infoToNotification(String message) {
        logToNotification(message, OMFLogLevel2.INFO);
    }

    public static void warnToUIConsole(String message) {
        logToUIConsole(message, OMFLogLevel2.WARNING);
    }

    public static void warnToNotification(OMFLog2 message) {
        logToNotification(message, OMFLogLevel2.WARNING);
    }

    public static void errorToNotification(OMFLog2 message) {
        logToNotification(message, OMFLogLevel2.ERROR);
    }

    public static void infoToNotification(OMFLog2 message) {
        logToNotification(message, OMFLogLevel2.INFO);
    }

    public static void warnToUIConsole(OMFLog2 message) {
        logToUIConsole(message, OMFLogLevel2.WARNING);
    }

    public static void errorToUIConsole(String message) {
        logToUIConsole(message, OMFLogLevel2.ERROR);
    }

    public static void errorToUIConsole(OMFLog2 message) {
        logToUIConsole(message, OMFLogLevel2.ERROR);
    }

    public static void infoToUIConsole(OMFLog2 message) {
        logToUIConsole(message, OMFLogLevel2.INFO);
    }

    public static void infoToUIConsole(String message) {
        logToUIConsole(message, OMFLogLevel2.INFO);
    }

    public static void warnToSystemConsole(String message) {
        logToSystemConsole(message, OMFLogLevel2.WARNING);
    }

    public static void errorToSystemConsole(String message) {
        logToSystemConsole(message, OMFLogLevel2.ERROR);
    }

    public static void infoToSystemConsole(String message) {
        logToSystemConsole(message, OMFLogLevel2.INFO);
    }

    public static void warnToSystemConsole(OMFLog2 message) {
        logToSystemConsole(message, OMFLogLevel2.WARNING);
    }

    public static void errorToSystemConsole(OMFLog2 message) {
        logToSystemConsole(message, OMFLogLevel2.ERROR);
    }

    public static void infoToSystemConsole(OMFLog2 message) {
        logToSystemConsole(message, OMFLogLevel2.INFO);
    }


    public static void defaultWarningException(OMFWarningException warningException) {
        warnToNotification(warningException.getUiMessage());
        warnToUIConsole(warningException.getUiMessage());
        warnToSystemConsole(warningException.getUiMessage());
    }
}

File: omf-core-framework/src/main/java/com/samares_engineering/omf/omf_core_framework/errormanagement2/logging/OMFLogger2.kt
package com.samares_engineering.omf.omf_core_framework.errormanagement2.logging

import com.google.common.base.Strings
import com.samares_engineering.omf.omf_core_framework.errormanagement2.exceptions.CoreException2
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.concrete_loggers.NotificationLogger
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.concrete_loggers.SystemLogger
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.concrete_loggers.UILogger
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.log.OMFColors
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.log.OMFLog
import com.samares_engineering.omf.omf_core_framework.errormanagement2.logging.log.OMFLogLevel
import com.samares_engineering.omf.omf_core_framework.feature.OMFFeature
import com.samares_engineering.omf.omf_core_framework.plugin.OMFPlugin

class OMFLogger2 private constructor(
    private val plugin: OMFPlugin
) {

    private val mapLogger: MutableMap<String, OMFConcreteLogger> = mutableMapOf(
        LogTarget.UI_CONSOLE.toString() to UILogger(plugin),
        LogTarget.SYSTEM_CONSOLE.toString() to SystemLogger(plugin),
        LogTarget.NOTIFICATION.toString() to NotificationLogger(plugin)
    )
    private val logLevel: OMFLogLevel = OMFLogLevel.INFO
    private var currentTarget: LogTarget? = null
    private var feature: OMFFeature? = null

    val uiLogger
        get() = mapLogger["UI_CONSOLE"]
    val systemLogger
        get() = mapLogger["SYSTEM_CONSOLE"]
    val notificationLogger
        get() = mapLogger["NOTIFICATION"]

    enum class LogTarget {
        UI_CONSOLE, SYSTEM_CONSOLE, NOTIFICATION, ALL
    }

    companion object {
        @JvmStatic
        private var instance: OMFLogger2? = null

        @JvmStatic
        fun init(plugin: OMFPlugin) {
            if (instance != null) {
                throw CoreException2("OMFLogger2 has already been initialized.")
            }
            instance = OMFLogger2(plugin)
        }

        @JvmStatic
        private fun getInstance(): OMFLogger2 {
            if (instance == null) {
                throw CoreException2("OMFLogger2 not initialized. Call init() first.")
            }
            return instance as OMFLogger2
        }

        @JvmStatic
        fun toUI(): OMFLogger2 {
            return getInstance().apply { currentTarget = LogTarget.UI_CONSOLE }
        }

        @JvmStatic
        fun toSystem(): OMFLogger2 {
            return getInstance().apply { currentTarget = LogTarget.SYSTEM_CONSOLE }
        }

        @JvmStatic
        fun toNotification(): OMFLogger2 {
            return getInstance().apply { currentTarget = LogTarget.NOTIFICATION }
        }

        @JvmStatic
        fun toAll(): OMFLogger2 {
            return getInstance().apply { currentTarget = LogTarget.ALL }
        }

        // Utility methods
        fun getPrefix(logLevel: OMFLogLevel): String {
            val pluginPrefix = if (instance?.plugin != null) "[${instance?.plugin?.name}]" else ""
            return "[" + getLogLevelPrefix(logLevel) + "]" + pluginPrefix
        }

        fun getPrefix(logLevel: OMFLogLevel, featureName: String): String {
            val featureTag = if (Strings.isNullOrEmpty(featureName)) "" else "[$featureName]"
            return getPrefix(logLevel) + featureTag
        }

        private fun getLogLevelPrefix(logLevel: OMFLogLevel): String {
            return when (logLevel) {
                OMFLogLevel.WARNING -> "Warning"
                OMFLogLevel.ERROR -> "Error"
                OMFLogLevel.INFO -> "Info"
                else -> "Info"
            }
        }
    }

    fun feature(feature: OMFFeature): OMFLogger2 {
        return getInstance().apply { this.feature = feature }
    }

    fun log(message: OMFLog) {
        log(message, OMFLogLevel.INFO)
    }

    fun success(message: String) {
        log(OMFLog().color(message, OMFColors.GREEN), OMFLogLevel.SUCCESS)
    }

    fun success(message: OMFLog) {
        log(message.colorAll(OMFColors.GREEN), OMFLogLevel.SUCCESS)
    }

    fun warning(message: String) {
        log(OMFLog().color(message, OMFColors.WARN), OMFLogLevel.WARNING)
    }

    fun warning(message: OMFLog) {
        log(message.colorAll(OMFColors.WARN), OMFLogLevel.WARNING)
    }

    fun error(message: String) {
        log(OMFLog(message).colorAll(OMFColors.ERROR), OMFLogLevel.ERROR)
    }

    fun error(message: OMFLog) {
        log(message.colorAll(OMFColors.ERROR), OMFLogLevel.ERROR)
    }

    private fun log(logMessage: OMFLog, logLevel: OMFLogLevel) {
        if (logLevel.ordinal >= this.logLevel.ordinal) {
            val formattedLog = logMessage.toHTMLFormat(logLevel, feature?.name)
            when (currentTarget) {
                LogTarget.UI_CONSOLE -> logToMDConsole(formattedLog, logMessage)
                LogTarget.NOTIFICATION -> logToNotification(logLevel, logMessage)
                LogTarget.SYSTEM_CONSOLE -> printToSystemConsole(logLevel, logMessage.toString())
                LogTarget.ALL -> {
                    logToMDConsole(formattedLog, logMessage)
                    logToNotification(logLevel, logMessage)
                    printToSystemConsole(logLevel, logMessage.toString())
                }

                null -> {}
            }
        }
    }

    private fun logToMDConsole(
        formattedLog: String?,
        logMessage: OMFLog
    ) {
        uiLogger?.log(formattedLog ?: "", logMessage, plugin, feature, logLevel)
    }

    private fun logToNotification(
        logLevel: OMFLogLevel,
        logMessage: OMFLog
    ) {
        notificationLogger?.log("", logMessage, plugin, feature, logLevel)
    }

    private fun printToSystemConsole(logLevel: OMFLogLevel, message: String) {
        val log = OMFLog(message)
        systemLogger?.log("", log, plugin, feature, logLevel)
    }

    @GenerateLegacyMethods
    class OMFLogger2LegacyMethods {

    }
}